openstack/baremetal/apiversions/results.go

```diff
@@ -0,0 +1,62 @@
+package apiversions
+
+import (
+	"github.com/gophercloud/gophercloud"
+)
+
+// APIVersions represents the result from getting a list of all versions available
+type APIVersions struct {
+	DefaultVersion APIVersion   `json:"default_version"`
+	Versions       []APIVersion `json:"versions"`
+}
+
+// APIVersion represents an API version for Ironic
+type APIVersion struct {
+	// ID is the unique identifier of the API version.
+	ID string `json:"id"`
+
+	// MinVersion is the minimum microversion supported.
+	MinVersion string `json:"min_version"`
```

Comment:
Do you maybe want to convert versions to a tuple (int, int) or any other useful type in Go?

Reply:
I was planning on using https://github.com/hashicorp/go-version in terraform-provider-ironic to compare versions.  Is Ironic guaranteed to only ever have X.Y microversions for all time, forever?

I'm not really sure I like the idea of enforcing a constraint in gophercloud on what the API returns for these values.

Reply:
>  Is Ironic guaranteed to only ever have X.Y microversions for all time, forever?

This is a definition of a microversion, so yes.

Reply:
Also: the other service that returns apiversions for its services uses strings (Manila) uses strings in gophercloud, so I think I need to be consistent.

Reply:
I don't think we should be consistent with suboptimal design, but up to you

Reply:
I disagree that it's suboptimal.  If it were optimal to return tuples of ints, why doesn't the API already do it?

Reply:
Because JSON doesn't have tuples :) I don't suggest naked tuples though, probably a smarter data type that has comparison and serialization back into X.Y. Everyone who'll ever use your new API will need either or both. If you pardon me a reference to rust, something like https://docs.rs/osauth/0.2.2/osauth/struct.ApiVersion.html.

Reply:
It has lists and ints, the API could very well return [1,52] instead of "1.52"  I'll look into this, but I still think when different services do things differently in gophercloud it becomes harder to use, not easier.

Reply:
> I disagree that it's suboptimal.

Agreed with the disagree.

Reply:
> I don't think we should be consistent with suboptimal design, but up to you

Before we start throwing stones, I encourage you to search gophercloud issues and PRs for the keyword "microservice" and read the numerous comments on how best to implement it. If you still feel it's suboptimal, please provide a more constructive suggestion.

Reply:
> read the numerous comments on how best to implement it

I literally wrote an OpenStack API SIG guidance on how to use microversions in SDK, suggesting me this is... interesting.

> please provide a more constructive suggestion

Well, I don't mind copying myself:

> I don't suggest naked tuples though, probably a smarter data type that has comparison and serialization back into X.Y. Everyone who'll ever use your new API will need either or both.

Reply:
@dtantsur Why is having the end user of gophercloud using something purpose built for this like `go-version` not acceptable? I don't see why I should have to re-invent the wheel here.

Reply:
Because essentially everyone using this new API will have to do it. An SDK should optimize for common cases. It's the same reason that you serialize the response to this structure rather than just returning a string with JSON

Reply:
> I literally wrote an OpenStack API SIG guidance on how to use microversions in SDK

This document? https://specs.openstack.org/openstack/api-sig/guidelines/sdk-exposing-microversions.html

If so, this is certainly a great document and thank you for taking the time to write it. However, it wasn't available two years ago when we began implementing microversions in Gophercloud.

> suggesting me this is... interesting.

Just to clarify, I meant how to best implement it in Gophercloud -- I was not implying that you need to go look at microversions in general 🙂 

I've read over all specs for Microversions and while I support microversions and what they're accomplishing, they are nearly impossible to implement in a static typed language in a manageable way. In addition, the specs and documents I've read all take the point of view of a perfect world where no one has to support things _prior_ to the document's recommendations - that adds another layer of difficulty.

If we're talking about _this_ specific issue of parsing a min and max version at the API endpoint, sure, the problem appears simple: we can either return a string and have the user do parsing on their own or we can create a more structured data type and implement that.

The problems we ran into at the time were more that there was inconsistency with API endpoint results between services and unclear use on what to do with the API endpoint results (ie we found it wouldn't be necessary to parse the min and max frequently, therefore, we leave that to the application to do).

The only guaranteed use of parsing a microversion we could find at the time was to:

1. Poll the api endpoint
2. Grab min and max
3. Use one as the value of `client.Microversion`.

All of which can be done with a string.

It wasn't until a few weeks ago that I personally ran into the first occurrence of having to compare min/max/given in a structured way. Not that having to do that isn't a valid action, but I'm just trying to illustrate the decisions that lead up to how things are.

Now, if something in Gophercloud appears to be "suboptimal", there could be two reasons:

1. It was an innocent mistake and not intentionally designed as suboptimal. 
2. There's history behind the decision.

With all of this in mind, knowing what I know about Gophercloud and quirks of the OpenStack API, my recommendation is to keep things as a string and create a shared convenience function in `gophercloud/utils` which can parse and compare microversions.

Reply:
> Because essentially everyone using this new API will have to do it.

I am not trying to disagree with you just to disagree - we seem to have a different point of view here.

Parsing the min and max is only required when one wants to see if a requested feature is valid with the client's set microversion. This only happens when the client is attempting to use a feature that deviates from the base API version.

In my experience, this isn't a frequent event that happens. The vast majority of the time, applications that interact with OpenStack are only using the base API features.

But that's not to say that comparing microversions isn't a valid action. This is why I'm suggesting to create a reusable function in `gophercloud/utils`.

Reply:
> This document? https://specs.openstack.org/openstack/api-sig/guidelines/sdk-exposing-microversions.html

Yep

> However, it wasn't available two years ago when we began implementing microversions in Gophercloud.

I know :( This came WAY too late, and the whole microversioning story was an example of how NOT to rush features in. We did all sorts of mistakes in python-ironicclient because of that, and I think gophercloud tries to repeat some of them.

> they are nearly impossible to implement in a static typed language in a manageable way

I could challenge that after having some experience with implementing Nova microversions in rust-openstack. I do agree that it's very painful, much more painful than with Python and co. I'm ready to provide any guidance on making this happen.

> it wouldn't be necessary to parse the min and max frequently

Then what's the goal if this patch? I assumed that this API is added so that you can reason about microversions. These are the common patterns:

1. Okay, a user is requesting a feature that is available starting with 1.11. Min is 1.1, max is 1.56. We're good with requesting 1.11.
2. The Node object is coded with features from 1.11 to 1.56 in mind. What do we have here? Max 1.58? No, I don't know that, let's stick with 1.56.
2.1. Oh, Max is only 1.42? Then use 1.42.

> Use one as the value of client.Microversion

This is not quite correct, as I showed in the examples above. And this is one of the subtle things we learned the hard way :( So I'm trying to warn you against taking such short-cuts.

Say, for example, min=1.1, max=1.56. Imagine gophercloud is coded with 1.42 in mind. Which version do you use (example 2.1 above)?

1. If you use 1.56, you may be broken, because you don't know what changed between 1.42 and 1.56.
2. If you use 1.1, you lose all non-default features from 1.2 to 1.42. It may not seem important, but as I illustrate below it may prevent you from even enrolling a usable node.

There has to be version negotiation somewhere, and I'm pretty sad that gophercloud is not doing it (I do recognize that the ship has probably sailed on this one).

> The vast majority of the time, applications that interact with OpenStack are only using the base API features.

It may be more or less the case for things like Nova which had a nearly complete API at the time they introduced microversioning (Kilo?). It's not the case for Ironic. I have three examples to illustrate it:

1. Enrolling a node the way modern Nova and/or Allocation API will recognize it requires populating `resource_class` field introduced in API 1.21. Using older versions it's simply impossible to make a node recognized as ready for deployment (and if you think it sucks, you're right, but, as in your case, there is a lot of history here).
2. Such a popular feature as inspection is also only available with non-default API version. It's not possible to trigger it otherwise.
3. You cannot even trigger initial cleaning, since the enrolled nodes with appear directly in the `available` state, and both `clean` and `manage` actions were not available in 1.1. Well, a correction: before 1.2 the `available` state was just JSON `null` (and I wonder if gophercloud takes this into account).

Reply:
Maybe APIVersion could add some methods to provide the sorts of tests @dtantsur describes and we could still keep the data type simple?

Reply:
This is a good compromise. Then we can also add similar methods to services that already use strings, keeping things consistent.

Reply:
Thank you for your reply - it's very much appreciated.

What I was trying to do in my last reply is walk you through how the decision was made to use `string`. I understand microversions and I understand all patterns you have described. But _at the time decisions were made_, there was only a certain amount of information available. It's a little offensive to have someone drop in and say it's all "suboptimal" without knowing the history.

> I could challenge that after having some experience with implementing Nova microversions in rust-openstack. I do agree that it's very painful, much more painful than with Python and co. I'm ready to provide any guidance on making this happen.

The offer is appreciated. Please see https://github.com/gophercloud/gophercloud/issues/441 for prior discussions about implementing microversions in Go. If you have any ideas beyond those already discussed that make managing microversions easier, you'd be a hero.

> These are the common patterns

One thing to keep in mind is that Gophercloud is a Language Binding (as described in your document). These kinds of patterns are arguably more suited for a higher level SDK (which, at this time, is `gophercloud/utils`, which is really just a repo of loosely collected helper functions. It may some day become a full-blown SDK).

> Imagine gophercloud is coded with 1.42 in mind

Gophercloud should not be coded for any specific OpenStack release or microversion. It should provide the foundational means to communicate with major OpenStack API versions.

It's up to the application that consumes Gophercloud to choose to support a specific version or microversion. If there are repeated patterns across applications, then these should be aggregated into `gophercloud/utils`.

> There has to be version negotiation somewhere, and I'm pretty sad that gophercloud is not doing it (I do recognize that the ship has probably sailed on this one).

The ship is still in the dock. Another area that is causing possible confusion is understanding that Gophercloud is still very much unfinished. Gophercloud has two people with commit rights and we both volunteer our spare time to work on this.

See https://github.com/gophercloud/gophercloud/issues/1520 for the most recent discussion about this topic.

As I mentioned before, if something seems "suboptimal", there are two reasons: it's a mistake or there's history. In this case, the history is that no one has gotten around to it yet. 

> It's not the case for Ironic.

I completely understand this and completely think this is valid.

But if you were to zoom out of Ironic and look at all projects under the OpenStack umbrella, the patterns you described aren't widely used (yet). 

Please understand all I was trying to do is provide a defensive reasoning for what you were calling "suboptimal". The implementation was done by looking at things in a completely different point of view which I will argue _is not incorrect_. But not being incorrect is different than _being the only solution_. I'm very much open to alternative implementations - I just appreciate if the person suggesting them takes time to look at why something has been implemented initially.

> Maybe APIVersion could add some methods to provide the sorts of tests @dtantsur describes and we could still keep the data type simple?

I think this might be similar to what I was talking about with adding some helper functions to `gophercloud/utils`.

Reply:
> Maybe APIVersion could add some methods to provide the sorts of tests @dtantsur describes and we could still keep the data type simple?

What do you mean by APIVersion here? APIVersion is a result type specific to Ironic in Gophercloud.  I don't think it makes sense to add methods to it as it'd only be useful to Ironic. Manila's APIVersion struct is different, because for whatever reason, openstack services don't seem to return exactly the same format for versioning.  We'd need to declare a general version type different than string to get methods that work everywhere.

@jtopjian proposal is putting helper functions in gophercloud/utils that can compare versions stored as strings.  That seems like a compromise, but I still don't really like it.  See below.

Reply:
> Then what's the goal if this patch? I assumed that this API is added so that you can reason about microversions. These are the common patterns:
> 
>     1. Okay, a user is requesting a feature that is available starting with 1.11. Min is 1.1, max is 1.56. We're good with requesting 1.11.
>
>     2. The Node object is coded with features from 1.11 to 1.56 in mind. What do we have here? Max 1.58? No, I don't know that, let's stick with 1.56.
>        2.1. Oh, Max is only 1.42? Then use 1.42.

The goal of the patch is to get the version data from the API, in the format the API gives it to us, and let the client do whatever they want with it. Gophercloud [leaves it up to the client](https://github.com/gophercloud/gophercloud/blob/master/docs/MICROVERSIONS.md#application-developer-information) to handle more advanced use cases. I understand the desire to want to make gophercloud do everything, and include all the kinds of helpers end users could possibly want.  This is similar to building the config drive ISO.  Well, that didn't end up in gophercloud.  It ended up in `gophercloud/utils` - because that's where quality of life features that help out with gophercloud go, **and, most importantly**, there wasn't already a good golang solution for it. In this case, **there is.**

In my case, my purpose is to reason about microversions in my terraform provider.  Let's say a user defines an allocation resource, this is similar to example 1 from @dtantsur above.  I already have a plan on how to handle this in my provider - using a library that already exists.

```golang
// Ironic server's range
ironicRange = fmt.Sprintf(">= %s =< %s", apiVersion.MinVersion, apiVersion.Version)
constraint, err := version.NewConstraint(ironicRange)
if err != nil {
    return err
}

// Minimum required for allocations
minimumVersion = "1.52"

// Check if server supports 1.52, and if so set our microversion to 1.52, otherwise
// we let the user know they can't use allocations with this endpoint.
if constraints.Check(minimumVersion) {
  client.Microversion = minimumVersion
} else
  return fmt.Errorf("Allocations are not supported by this Ironic server.")
}
```

Why do I need gophercloud to give me anything other than a string? Why are you so determined to have me rewrite a library and incorporate the functionality that exists in `go-version` here in gophercloud?

Reply:
> The ship is still in the dock. Another area that is causing possible confusion is understanding that Gophercloud is still very much unfinished. Gophercloud has two people with commit rights and we both volunteer our spare time to work on this.

Thank you for that. I've found the code and its maintainers all quite nice to work with. :-)

> > Maybe APIVersion could add some methods to provide the sorts of tests @dtantsur describes and we could still keep the data type simple?
> 
> I think this might be similar to what I was talking about with adding some helper functions to `gophercloud/utils`.

Yeah, I think the distinction is just geographical. If the functions are on APIVersion, then a consumer doesn't have to import helpers from somewhere else. OTOH, as you point out, operating on microversion ranges could be considered a higher-level SDK-like feature, which would make more sense in the utils repo. I'm OK with either approach, myself, but was tossing the idea out there as a way to bridge the two proposals being discussed here.

Reply:
> > Maybe APIVersion could add some methods to provide the sorts of tests @dtantsur describes and we could still keep the data type simple?
> 
> What do you mean by APIVersion here? APIVersion is a result type specific to Ironic in Gophercloud. I don't think it makes sense to add methods to it as it'd only be useful to Ironic. Manila's APIVersion struct is different, because for whatever reason, openstack services don't seem to return exactly the same format for versioning. 

Of course not, why would any 2 things coming out of OpenStack do the same thing in the same way?! ;-)

Reply:
> In my case, my purpose is to reason about microversions in my terraform provider. Let's say a user defines an allocation resource, this is similar to example 1 from @dtantsur above. I already have a plan on how to handle this in my provider - using a library that already exists.
> 
> ```go
> // Ironic server's range
> ironicRange = fmt.Sprintf(">= %s, <= %s", apiVersion.MinVersion, apiVersion.Version)
> constraint, err := version.NewConstraint(ironicRange)
> if err != nil {
>     return err
> }
> 
> // Minimum required for allocations
> minimumVersion = "1.52"
> 
> // Check if server supports 1.52, and if so set our microversion to 1.52, otherwise
> // we let the user know they can't use allocations with this endpoint.
> if constraints.Check(minimumVersion) {
>   client.Microversion = minimumVersion
> } else
>   return fmt.Errorf("Allocations are not supported by this Ironic server.")
> }
> ```
> 
> Why do I need gophercloud to give me anything other than a string? Why are you so determined to have me rewrite a library and incorporate the functionality that exists in `go-version` here in gophercloud?

The approach you have above looks good to me (I wasn't aware of go-version).

Reply:
>  It's a little offensive to have someone drop in and say it's all "suboptimal" without knowing the history.
> The implementation was done by looking at things in a completely different point of view which I will argue is not incorrect.

Sorry if it looked this way. However, you may be reading too much in my words, all I meant by "suboptimal" is "not the best way" (and I've just double-checked my English with Merriam-Webster, it seems to agree). I never meant any offence to the gophercloud team, nor did I mean "incorrect".

> If you have any ideas beyond those already discussed that make managing microversions easier, you'd be a hero.

I'll try to, although my lack of familiarity with Go can be problematic. On the other hand, I have a TODO item to write some text on how we treat microversions in python-openstacksdk..

> One thing to keep in mind is that Gophercloud is a Language Binding

The general stance right now is "please write SDKs, not language bindings", although I do know how much harder it is. The problem is here:

> Gophercloud should not be coded for any specific OpenStack release or microversion.

It kind of is. The structures you provide are valid for a certain microversion range, usually at the time of contributing them. They may get broken in future microversions. If not enough testing was done (and testing is one of the weakest points of microversioning), some old versions may not work either (e.g. I cannot quite figure out if we handle the case of Node.provision_state == null).

> It's up to the application that consumes Gophercloud to choose to support a specific version or microversion

This is a fair call for a dynamically typed language, but you have less choice with something like Go. You can provide any microversion, but you cannot update the structures (e.g. Node) to contain the necessary fields or CreateOpts to contain new arguments.

On the other hand, you have enough information to make judgement. E.g. you see that Node's creation request contains `resource_class != null`. Okay, now you know that your minimum microversion is 1.21, because older microversions cannot satisfy such request. However, if you support newer Node fields (and you do, up to IIRC 1.56), you can choose to figure out a higher version. E.g. server supports up to 1.58, but you don't implement anything beyond 1.56. You pick 1.56 in the end. This is pretty much a TL;DR, as I said, I'm tasked with writing a more organized text.

Again, this all was a late understanding. I'm not trying to blame anyone. We all did it wrong initially (and FWIW I've never been a proponent of microversioning).

Reply:
My main (and probably only) point here @stbenjam is that the code you've written as an example should be in gophercloud, not in every consumer of gophercloud. As not doing it seems to be a weighted decision, I'm not going to insist.

Reply:
> I never meant any offence to the gophercloud team, nor did I mean "incorrect".

The apology should be all mine - I'm usually more mindful of these kinds of interpretations and I failed at this one.

> Thank you for that. I've found the code and its maintainers all quite nice to work with. :-)

Thanks, Doug!

> but was tossing the idea out there as a way to bridge the two proposals being discussed here.

I didn't mention this before, but this is how I interpreted it and it's appreciated.

> The approach you have above looks good to me 

+1

>  there wasn't already a good golang solution for it. In this case, there is.

This is a good point and deserves to be elaborated on. 

In Go, it is acceptable to duplicate common blocks of code across projects rather than centralize them in a common library. 

If @stbenjam wants to implement something with `go-version`, go for it. Slightly different from @dhellmann, I knew of `go-version`, but didn't have it in mind when I wrote [this](https://github.com/terraform-providers/terraform-provider-openstack/blob/master/openstack/util.go#L278) in Terraform OpenStack. There's also an implementation in [cloud provider](https://github.com/kubernetes/cloud-provider-openstack/blob/master/pkg/share/manila/client.go#L140).

Maybe these three implementations could be aggregated into a common `gophercloud/utils` function. Maybe not. I don't want to force that without seeing how things naturally unfold. This approach can delay development, though. If I was able to work full-time on Gophercloud, I might take a more directed approach.

I'd like to try to wrap up this discussion because these things can get unwieldy in Github.

For now, let's go with string, but I am absolutely not opposed to making changes. If that's something to investigate, we can open a dedicated issue for this and continue discussion there. I understand all points of view here and think they are all valid.